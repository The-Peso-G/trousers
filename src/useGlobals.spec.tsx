import React from 'react';
import { render, cleanup } from '@testing-library/react';

import {
    css,
    useGlobals,
    SingleStyleCollector,
    ThemeProvider,
    useStyles,
    styleCollector,
} from './';

interface Theme {
    backgroundColor: string;
}

interface FixtureProps {
    styles: SingleStyleCollector<Theme> | SingleStyleCollector<Theme>[];
}

describe('useGlobals', () => {
    let theme: Theme;
    let globalStyles: SingleStyleCollector<Theme>;
    let GlobalyStyledComponent: React.FC<FixtureProps>;

    beforeEach(() => {
        globalStyles = css`
            * {
                box-sizing: border-box;
            }
        `;

        theme = {
            backgroundColor: 'red',
        };

        GlobalyStyledComponent = ({ styles }: FixtureProps) => {
            useGlobals(styles);

            return <span>Ahoy!</span>;
        };
    });

    afterEach(() => {
        document.getElementsByTagName('html')[0].innerHTML = '';

        cleanup();
    });

    it('attaches styles to the head', () => {
        render(<GlobalyStyledComponent styles={globalStyles} />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches an array of styles to the head', () => {
        const additionalStyles = css`
            * {
                background-color: blue;
            }
        `;

        render(
            <GlobalyStyledComponent
                styles={[globalStyles, additionalStyles]}
            />,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches new styles when a new style collector is supplied', () => {
        const additionalStyles = css`
            * {
                background-color: blue;
            }
        `;

        const { rerender } = render(
            <GlobalyStyledComponent styles={globalStyles} />,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        rerender(<GlobalyStyledComponent styles={additionalStyles} />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('styles are not duplicated on rerender', () => {
        const { rerender } = render(
            <GlobalyStyledComponent styles={globalStyles} />,
        );
        rerender(<GlobalyStyledComponent styles={globalStyles} />);
        rerender(<GlobalyStyledComponent styles={globalStyles} />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches themed styles to the head', () => {
        const themedStyles = css<Theme>`
            * {
                background-color: ${theme => theme.backgroundColor};
            }
        `;

        render(
            <ThemeProvider theme={theme}>
                <GlobalyStyledComponent styles={themedStyles} />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches an array of themed styles to the head', () => {
        const themedStyles = css<Theme>`
            * {
                background-color: ${theme => theme.backgroundColor};
            }
        `;

        render(
            <ThemeProvider theme={theme}>
                <GlobalyStyledComponent styles={[globalStyles, themedStyles]} />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches styles before data-trousers', () => {
        const styles = styleCollector('hello').element`
            color: blue;
        `;

        const MyComponent = () => {
            const classNames = useStyles(styles);

            return <div className={classNames}>Hello!</div>;
        };

        render(
            <React.Fragment>
                <MyComponent />
                <GlobalyStyledComponent styles={globalStyles} />
            </React.Fragment>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('clears styles', () => {
        const { unmount } = render(
            <GlobalyStyledComponent styles={globalStyles} />,
        );

        expect(document.querySelectorAll('style').length).toEqual(1);

        unmount();

        expect(document.querySelectorAll('style').length).toEqual(0);
    });

    it('clears an array of styles', () => {
        const additionalStyles = css`
            * {
                background-color: blue;
            }
        `;

        const { unmount } = render(
            <GlobalyStyledComponent
                styles={[globalStyles, additionalStyles]}
            />,
        );

        expect(document.querySelectorAll('style').length).toEqual(1);

        unmount();

        expect(document.querySelectorAll('style').length).toEqual(0);
    });

    it('clears altered styles', () => {
        const additionalStyles = css`
            * {
                background-color: blue;
            }
        `;

        const { rerender, unmount } = render(
            <GlobalyStyledComponent styles={globalStyles} />,
        );

        expect(document.querySelectorAll('style').length).toEqual(1);

        rerender(<GlobalyStyledComponent styles={additionalStyles} />);

        unmount();

        expect(document.querySelectorAll('style').length).toEqual(0);
    });
});
