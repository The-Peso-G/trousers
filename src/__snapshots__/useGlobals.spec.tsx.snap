// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGlobals attaches an array of styles to the head 1`] = `
<head>
  <style
    data-trousers-global="436679506 1211230107"
    type="text/css"
  >
    *{box-sizing:border-box;}

    *{background-color:blue;}

  </style>
</head>
`;

exports[`useGlobals attaches an array of themed styles to the head 1`] = `
<head>
  <style
    data-trousers-global="436679506 664824792"
    type="text/css"
  >
    *{box-sizing:border-box;}

    *{background-color:red;}

  </style>
</head>
`;

exports[`useGlobals attaches new styles when a new style collector is supplied 1`] = `
<head>
  <style
    data-trousers-global="436679506"
    type="text/css"
  >
    *{box-sizing:border-box;}

  </style>
</head>
`;

exports[`useGlobals attaches new styles when a new style collector is supplied 2`] = `
<head>
  <style
    data-trousers-global="1211230107"
    type="text/css"
  >
    *{background-color:blue;}

  </style>
</head>
`;

exports[`useGlobals attaches styles before data-trousers 1`] = `
<head>
  <style
    data-trousers-global="436679506"
    type="text/css"
  >
    *{box-sizing:border-box;}

  </style>
  <style
    data-trousers=".hello__11148054330"
    type="text/css"
  >
    .hello__11148054330{color:blue;}

  </style>
</head>
`;

exports[`useGlobals attaches styles to the head 1`] = `
<head>
  <style
    data-trousers-global="436679506"
    type="text/css"
  >
    *{box-sizing:border-box;}

  </style>
</head>
`;

exports[`useGlobals attaches themed styles to the head 1`] = `
<head>
  <style
    data-trousers-global="664824792"
    type="text/css"
  >
    *{background-color:red;}

  </style>
</head>
`;

exports[`useGlobals styles are not duplicated on rerender 1`] = `
<head>
  <style
    data-trousers-global="436679506"
    type="text/css"
  >
    *{box-sizing:border-box;}

  </style>
</head>
`;
